syntax = "proto3";

import "google/protobuf/empty.proto";

package master;

// A task is a unit of work for ElasticDL training workers, assigned by master.
// Worker divides a task into multiple minibatches and compute a gradient for
// each minibatch. For now, only RecordIO file format is supported.
message Task {
    int32 minibatch_size = 1;

    // Filename for a RecordIO shard. An empty filename signifies that the
    // master has no pending taske to assign to requesting worker.
    string shard_file_name = 2;

    // Strating and ending (inclusive) record number.
    int64 start = 3;
    int64 end = 4;

    // Current model version of master
    int32 model_version = 5;
}

// Modeled after
// https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.proto
// but without external dependencies.
message Tensor {
    enum DataType {
        DT_INVALID = 0;
        DT_FLOAT = 1;
        DT_DOUBLE = 2;
    }
    DataType dtype = 1;
    
    // Dimensions of the tensor. The first entry in "dim" is the outermost
    // dimension used to layout the values, the last entry is the innermost
    // dimension.
    repeated int32 dim = 2;

    // ndarray's buffer dump.
    bytes content = 3;
}

message Model {
    int32 version = 1;
    map<string, Tensor> param = 2; 
}

message GetModelRequest {
    int32 min_version = 1;
}

service Master {
    rpc GetTask(google.protobuf.Empty) returns (Task);
    rpc GetModel(GetModelRequest) returns (Model);
}
